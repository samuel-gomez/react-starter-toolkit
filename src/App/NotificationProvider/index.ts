export { default, NotificationContext } from './NotificationProvider';
export { default as useNotifications } from './Notifications.hook';
export type { TReturnUseNotification } from './Notifications.hook';
export type { TaddNotification, TonDeleteNotification } from './Notifications.hook';
export type { EType } from './Notifications';
