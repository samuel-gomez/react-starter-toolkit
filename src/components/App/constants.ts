import { ADMIN } from 'shared/constants';

// these mock functions are used when you disable the OIDC authentication
export const useOidcUserFnMock = () => ({ oidcUser: { name: 'John Doe', member_of: [ADMIN] } });
export const useOidcAccessTokenFnMock = () => ({ accessToken: 'token' });
