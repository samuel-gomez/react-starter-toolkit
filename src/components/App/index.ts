export { default } from './App.container';
export type { TApp } from './App';
