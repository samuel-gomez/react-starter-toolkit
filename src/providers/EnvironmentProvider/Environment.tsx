import { useState, useEffect, createContext, Dispatch, SetStateAction } from 'react';

export type TEnvironment = {
  apiUrl: Record<string, string>;
  fetchConfig: object;
  oidc: Record<string, boolean | string>;
};

export type TEnvironmentState = {
  environment?: null | TEnvironment;
  error?: unknown;
};

const initState: TEnvironmentState = {
  environment: null,
  error: null,
};

export const EnvironmentContext = createContext<TEnvironmentState>(initState);
EnvironmentContext.displayName = 'EnvironmentContext';

export const getFileEnv = (nodeEnv = process.env.REACT_APP_NODE_ENV || process.env.NODE_ENV) => `environment.${nodeEnv}.json`;

type TfetchEnv = {
  setEnvState: Dispatch<SetStateAction<TEnvironmentState>>;
  signal: AbortSignal;
  fetchFn?: typeof fetch;
  getFileEnvFn?: typeof getFileEnv;
};

export const fetchEnv = async ({ setEnvState, signal, fetchFn = fetch, getFileEnvFn = getFileEnv }: TfetchEnv) => {
  const fileName = getFileEnvFn();
  try {
    const response = await fetchFn(`/${fileName}`, { signal });
    const environment = await response?.json();
    setEnvState({ environment });
  } catch (error) {
    setEnvState({ error, environment: null });
  }
};

/**
 * Hook to setState environment
 */
export const useEnv = (fetchEnvFn = fetchEnv, initStateCt = initState) => {
  const [envState, setEnvState] = useState(initStateCt);
  useEffect(() => {
    const abortController = new AbortController();
    if (envState.environment === null) {
      fetchEnvFn({ setEnvState, signal: abortController.signal });
    }
    return () => abortController.abort();
  }, [envState.environment, fetchEnvFn]);

  return { envState, setEnvState };
};

/**
 * Component wrapper to add environment context
 * @param {Component} children
 */
const EnvironmentProvider = ({ children, useEnvFn = useEnv }: { children: JSX.Element; useEnvFn?: typeof useEnv }) => {
  const { envState } = useEnvFn();
  return <EnvironmentContext.Provider value={envState}>{children}</EnvironmentContext.Provider>;
};

export default EnvironmentProvider;
