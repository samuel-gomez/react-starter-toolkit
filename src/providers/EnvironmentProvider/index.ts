export { default, EnvironmentContext, useEnv, fetchEnv, getFileEnv } from './Environment';
export type { TEnvironment, TEnvironmentState } from './Environment';
