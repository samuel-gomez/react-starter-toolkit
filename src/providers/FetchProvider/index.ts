export { default, FetchContext } from './FetchProvider';
export type { TResponse } from './FetchProvider';
export { default as setResponseError } from './setResponseError';
