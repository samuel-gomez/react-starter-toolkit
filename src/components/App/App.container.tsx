'use client';
import { useContext } from 'react';
import Loader, { MODES } from 'shared/components/Loader';
import EnvironmentProvider, { EnvironmentContext } from 'providers/EnvironmentProvider';
import App from './App';

type TAppContainer = { children: React.ReactNode; EnvironmentContextObj?: typeof EnvironmentContext; AppCmpt?: typeof App };

export const AppContainer = ({ children, EnvironmentContextObj = EnvironmentContext, AppCmpt = App }: TAppContainer) => {
  const { environment } = useContext(EnvironmentContextObj);
  return !environment ? (
    <Loader text="Chargement de l'environnement..." mode={MODES.get} classModifier="fullscreen" />
  ) : (
    <AppCmpt {...environment}>{children}</AppCmpt>
  );
};

const AppWithEnvironment = ({ children }: { children: React.ReactNode }) => (
  <EnvironmentProvider>
    <AppContainer>{children}</AppContainer>
  </EnvironmentProvider>
);

export default AppWithEnvironment;
