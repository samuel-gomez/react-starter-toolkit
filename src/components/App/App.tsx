import { ReactNode } from 'react';
import { OidcProvider, useOidcUser, useOidcAccessToken, OidcSecure } from '@axa-fr/react-oidc';
import UserProvider from 'providers/UserProvider';
import FetchProvider from 'providers/FetchProvider';
import NotificationProvider from 'providers/NotificationProvider';
import type { TEnvironment } from 'providers/EnvironmentProvider';
import { useOidcAccessTokenFnMock, useOidcUserFnMock } from './constants';

type TUserAndFetchProviders = Omit<TEnvironment, 'oidc'> & {
  isEnabled: boolean;
  UserProviderCmpt: typeof UserProvider;
  FetchProviderCmpt: typeof FetchProvider;
  useOidcUserFn: typeof useOidcUser;
  useOidcAccessTokenFn: typeof useOidcAccessToken;
  children: ReactNode;
};

const UserAndFetchProviders = ({
  apiUrl,
  fetchConfig,
  isEnabled,
  useOidcUserFn,
  useOidcAccessTokenFn,
  UserProviderCmpt,
  FetchProviderCmpt,
  children,
}: TUserAndFetchProviders) => (
  <UserProviderCmpt isEnabled={isEnabled} useOidcUserFn={useOidcUserFn}>
    <FetchProviderCmpt apiUrl={apiUrl} fetchConfig={fetchConfig} useOidcAccessTokenFn={useOidcAccessTokenFn}>
      {children}
    </FetchProviderCmpt>
  </UserProviderCmpt>
);

type TOidc = Omit<TUserAndFetchProviders, 'isEnabled'> &
  TEnvironment & {
    children: ReactNode;
    OidcProviderCmpt: typeof OidcProvider;
    OidcSecureCmpt: typeof OidcSecure;
    useOidcUserFn: typeof useOidcUser;
    useOidcAccessTokenFn: typeof useOidcAccessToken;
  };

const Oidc = ({ oidc, OidcProviderCmpt, OidcSecureCmpt, ...restProps }: TOidc) =>
  oidc.isEnabled ? (
    <OidcProviderCmpt configuration={{ ...oidc }}>
      <OidcSecureCmpt>
        <UserAndFetchProviders {...restProps} isEnabled={Boolean(oidc.isEnabled)} />
      </OidcSecureCmpt>
    </OidcProviderCmpt>
  ) : (
    <UserAndFetchProviders
      {...restProps}
      isEnabled={Boolean(oidc.isEnabled)}
      useOidcUserFn={useOidcUserFnMock}
      useOidcAccessTokenFn={useOidcAccessTokenFnMock}
    />
  );

export type TApp = TEnvironment & {
  children: React.ReactNode;
  OidcCmpt?: typeof Oidc;
  useOidcUserFn?: typeof useOidcUser;
  useOidcAccessTokenFn?: typeof useOidcAccessToken;
  OidcProviderCmpt?: typeof OidcProvider;
  OidcSecureCmpt?: typeof OidcSecure;
  UserProviderCmpt?: typeof UserProvider;
  FetchProviderCmpt?: typeof FetchProvider;
};

const App = ({
  children,
  oidc,
  fetchConfig,
  apiUrl,
  OidcCmpt = Oidc,
  useOidcUserFn = useOidcUser,
  useOidcAccessTokenFn = useOidcAccessToken,
  OidcProviderCmpt = OidcProvider,
  OidcSecureCmpt = OidcSecure,
  UserProviderCmpt = UserProvider,
  FetchProviderCmpt = FetchProvider,
}: TApp) => (
  <OidcCmpt
    OidcProviderCmpt={OidcProviderCmpt}
    OidcSecureCmpt={OidcSecureCmpt}
    UserProviderCmpt={UserProviderCmpt}
    FetchProviderCmpt={FetchProviderCmpt}
    fetchConfig={fetchConfig}
    apiUrl={apiUrl}
    oidc={oidc}
    useOidcUserFn={useOidcUserFn}
    useOidcAccessTokenFn={useOidcAccessTokenFn}
  >
    <NotificationProvider>{children}</NotificationProvider>
  </OidcCmpt>
);

export default App;
