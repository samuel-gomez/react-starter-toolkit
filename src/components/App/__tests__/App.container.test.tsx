import { createContext } from 'react';
import { render, act } from '@testing-library/react';
import { TEnvironmentState } from 'providers/EnvironmentProvider';
import AppWithEnvironment, { AppContainer } from '../App.container';
import { TApp } from '../App';

describe('<AppWithEnvironment />', () => {
  it('Renders AppWithEnvironment component without crashing', async () => {
    const { asFragment } = render(<AppWithEnvironment>test</AppWithEnvironment>);
    await act(() => expect(asFragment()).toMatchSnapshot());
  });
});

describe('<AppContainer />', () => {
  it('Renders AppContainer component without crashing', () => {
    const API_URL = {
      base: '/myApiUrlEnv',
    };
    const EnvironmentContextObj = createContext<TEnvironmentState>({
      environment: {
        oidc: {},
        fetchConfig: {},
        apiUrl: API_URL,
      },
      error: null,
    });

    const AppCmpt = ({ apiUrl = API_URL }: TApp) => <p>{apiUrl.base}</p>;
    const { getByText } = render(
      <AppContainer EnvironmentContextObj={EnvironmentContextObj} AppCmpt={AppCmpt}>
        test
      </AppContainer>,
    );
    expect(getByText(/myApiUrlEnv/)).toBeInTheDocument();
  });
});
