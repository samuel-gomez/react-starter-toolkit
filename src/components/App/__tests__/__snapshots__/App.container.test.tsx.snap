// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppWithEnvironment /> Renders AppWithEnvironment component without crashing 1`] = `
<DocumentFragment>
  <div
    class="af-loader af-loader--fullscreen af-loader--active"
  >
    <p
      class="af-loader__spinner"
    >
      Chargement de l'environnement...
    </p>
  </div>
</DocumentFragment>
`;
